service/src/main/java/bio/terra/workspace/service/petserviceaccount/PetSaService.java

```diff
@@ -0,0 +1,254 @@
+              .getIamCow()
+              .projects()
+              .serviceAccounts()
+              .getIamPolicy(petServiceAccount)
+              .execute();
+      // If the caller supplied a non-null eTag value, it must match the current policy in order to
+      // modify the policy here. Otherwise, return without modifying the policy to prevent
+      // clobbering other changes.
+      if (eTag != null && !saPolicy.getEtag().equals(eTag)) {
+        logger.warn(
+            "GCP IAM policy eTag did not match expected value when revoking pet SA access for user {} in workspace {}.",
+            userAndPet.getUserEmail(),
+            workspaceId);
+        return Optional.empty();
+      }
+      // TODO(PF-991): when enablePetServiceAccountImpersonation stops putting the pet SA in this
+      //  binding, this method should stop removing it.
+      Binding bindingToRemove =
+          new Binding()
+              .setRole(SERVICE_ACCOUNT_USER_ROLE)
+              .setMembers(
+                  ImmutableList.of(
+                      "user:" + userAndPet.getUserEmail(),
+                      "serviceAccount:" + petServiceAccount.email()));
+      // If no bindings exist, getBindings() returns null instead of an empty list. If there are
+      // no policies, there is nothing to revoke, so this method is finished.
+      List<Binding> oldBindingList = saPolicy.getBindings();
+      if (oldBindingList == null) {
+        return Optional.empty();
+      }
+      List<Binding> newBindingsList = removeBinding(oldBindingList, bindingToRemove);
+      saPolicy.setBindings(newBindingsList);
+      SetIamPolicyRequest request = new SetIamPolicyRequest().setPolicy(saPolicy);
+      return Optional.of(
+          crlService
+              .getIamCow()
+              .projects()
+              .serviceAccounts()
+              .setIamPolicy(petServiceAccount, request)
+              .execute());
+    } catch (IOException e) {
+      throw new InternalServerErrorException("Error disabling user's pet SA", e);
+    }
+  }
+
+  /**
+   * Utility function for removing all instances of a GCP IAM binding from a list if it is present.
+   * List.remove does not work here as GCP may re-order the binding's member list.
+   */
+  private List<Binding> removeBinding(List<Binding> bindingList, Binding bindingToRemove) {
+    return bindingList.stream()
+        .filter(binding -> !bindingEquals(binding, bindingToRemove))
+        .collect(Collectors.toList());
+  }
+
+  private static boolean bindingEquals(Binding binding1, Binding binding2) {
+    return binding1.getRole().equals(binding2.getRole())
+        && binding1.getMembers().containsAll(binding2.getMembers())
+        && binding2.getMembers().containsAll(binding1.getMembers());
+  }
+
+  /**
+   * Returns the pet service account of a provided user in a provided project if it exists in GCP,
+   * or an empty Optional if it does not.
+   */
+  public Optional<ServiceAccountName> getArbitrarySa(
```

Comment:
nit: SAM has an endpoint to "gets a key for an arbitrary pet service account for the user", where "arbitrary" means any one of the pet SAs for that user. I think you mean "arbitrary" here in the sense that you want to lookup a SA without having the user's token available. Perhaps use a different word here to avoid confusion?